# LinkR - plugin Prescription

In [2]:
# Librairies:

import pandas as pd
import seaborn as sns
import matplotlib as plt
import plotly.io as pio
import plotly.express as px
import plotly.graph_objs as go
from plotly.subplots import make_subplots


# Drug Exposure

In [3]:
drug_exposure = pd.read_csv('Data/drug_exposure.csv')
print(drug_exposure.shape) #(18229, 23)
'''
['drug_exposure_id', 'person_id', 'drug_concept_id',
       'drug_exposure_start_date', 'drug_exposure_start_datetime',
       'drug_exposure_end_date', 'drug_exposure_end_datetime',
       'verbatim_end_date', 'drug_type_concept_id', 'stop_reason', 'refills',
       'quantity', 'days_supply', 'sig', 'route_concept_id', 'lot_number',
       'provider_id', 'visit_occurrence_id', 'visit_detail_id',
       'drug_source_value', 'drug_source_concept_id', 'route_source_value',
       'dose_unit_source_value']
'''
drug_exposure.person_id.value_counts()
'''
person_id
 7155255168997124770    1182
 4783904755296699562     791
 8090044958540695372     716
-3908355835367628651     687
 7131048714591189903     644
                        ... 
-2575767131279873665      37
-4183220989401122518      34
 1740609625029317924      26
 3912882389848878631      24
-8769042030325953499      18
'''
drug_exposure.info()
'''
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18229 entries, 0 to 18228
Data columns (total 23 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   drug_exposure_id              18229 non-null  int64  
 1   person_id                     18229 non-null  int64  
 2   drug_concept_id               18229 non-null  int64  
 3   drug_exposure_start_date      18229 non-null  object 
 4   drug_exposure_start_datetime  18229 non-null  object 
 5   drug_exposure_end_date        18229 non-null  object 
 6   drug_exposure_end_datetime    18229 non-null  object 
 7   verbatim_end_date             0 non-null      float64
 8   drug_type_concept_id          18229 non-null  int64  
 9   stop_reason                   0 non-null      float64
 10  refills                       0 non-null      float64
 11  quantity                      18220 non-null  float64
 12  days_supply                   0 non-null      float64
 13  sig                           0 non-null      float64
 14  route_concept_id              18229 non-null  int64  
 15  lot_number                    0 non-null      float64
 16  provider_id                   0 non-null      float64
 17  visit_occurrence_id           18229 non-null  int64  
 18  visit_detail_id               0 non-null      float64
 19  drug_source_value             18229 non-null  object 
 20  drug_source_concept_id        18229 non-null  int64  
 21  route_source_value            18223 non-null  object 
 22  dose_unit_source_value        18220 non-null  object 
 '''

drug_exposure.head()

liste_non_null = ['drug_exposure_id', 'person_id', 'drug_concept_id',  'drug_exposure_start_date', 'drug_exposure_start_datetime',
       'drug_exposure_end_date', 'drug_exposure_end_datetime', 'drug_type_concept_id',   'quantity',
       'route_concept_id', 'visit_occurrence_id', 'drug_source_value', 'drug_source_concept_id', 'route_source_value', 'dose_unit_source_value']


(18229, 23)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18229 entries, 0 to 18228
Data columns (total 23 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   drug_exposure_id              18229 non-null  int64  
 1   person_id                     18229 non-null  int64  
 2   drug_concept_id               18229 non-null  int64  
 3   drug_exposure_start_date      18229 non-null  object 
 4   drug_exposure_start_datetime  18229 non-null  object 
 5   drug_exposure_end_date        18229 non-null  object 
 6   drug_exposure_end_datetime    18229 non-null  object 
 7   verbatim_end_date             0 non-null      float64
 8   drug_type_concept_id          18229 non-null  int64  
 9   stop_reason                   0 non-null      float64
 10  refills                       0 non-null      float64
 11  quantity                      18220 non-null  float64
 12  days_supply                   0 non-null      fl

In [ ]:
drug_exposure.describe()


In [17]:
drug_exposure.drug_type_concept_id

0        32838
1        32838
2        32838
3        32838
4        32838
         ...  
18224    32838
18225    32838
18226    32838
18227    32838
18228    32838
Name: drug_type_concept_id, Length: 18229, dtype: int64

### Patient 1: 3912882389848878631
On se concentre sur un premier cas avec 24 lignes pour 1 personne: 
Attention, c'est un integer, pas un texte. Ne pas mettre de ''


In [4]:
patient1 = drug_exposure.loc[drug_exposure['person_id'] == 3912882389848878631, liste_non_null]
'''
patient1.columns
Index(['drug_exposure_id', 'person_id', 'drug_concept_id',
       'drug_exposure_start_date', 'drug_exposure_start_datetime',
       'drug_exposure_end_date', 'drug_exposure_end_datetime',
       'drug_type_concept_id', 'quantity', 'route_concept_id',
       'visit_occurrence_id', 'drug_source_value', 'drug_source_concept_id',
       'route_source_value', 'dose_unit_source_value'],
      dtype='object')
'''
patient1

drug_exposure_id            person_id  drug_concept_id  \
5622  6996991851384754967  3912882389848878631         40232756   
5623 -2528282390097198867  3912882389848878631         40232756   
5624  8598302712376844332  3912882389848878631         40232756   
5625  4557254971515693852  3912882389848878631         40232756   
5626  -290668764010704253  3912882389848878631         40221329   
5627  1423514582191454168  3912882389848878631         36249736   
5628   700275630367800584  3912882389848878631         19126251   
5629  -345527215447771312  3912882389848878631         19126251   
5630  -992298387767609339  3912882389848878631         19126251   
5631  6012719603598400287  3912882389848878631         19126251   
5632  7845585426484018192  3912882389848878631           734281   
5633 -3065974858089804864  3912882389848878631           734281   
5634 -2645922204242395989  3912882389848878631         19079250   
5635 -6600566077470540957  3912882389848878631         21125875   
5636  8139119374042756604  3912882389848878631         19137328   
5637 -3408295941477030482  3912882389848878631         19019418   
5638 -3343601482655155228  3912882389848878631         45775137   
5639 -1076779580276771548  3912882389848878631          1127433   
5640 -4804305780328125843  3912882389848878631           723020   
5641 -2002420539696298274  3912882389848878631         19077513   
5642 -6958167977848852902  3912882389848878631         43011850   
5643  8859302285208559102  3912882389848878631         19019050   
5644 -8668154338303612643  3912882389848878631         19127213   
5645  8330931891712023563  3912882389848878631         19127213   

     drug_exposure_start_date drug_exposure_start_datetime  \
5622               2129-01-05          2129-01-05 00:00:00   
5623               2129-01-04          2129-01-04 18:00:00   
5624               2129-01-05          2129-01-05 02:00:00   
5625               2129-01-04          2129-01-04 21:00:00   
5626               2129-01-05          2129-01-05 02:00:00   
5627               2129-01-05          2129-01-05 03:00:00   
5628               2129-01-10          2129-01-10 08:00:00   
5629               2129-01-11          2129-01-11 08:00:00   
5630               2129-01-08          2129-01-08 08:00:00   
5631               2129-01-06          2129-01-06 08:00:00   
5632               2129-01-05          2129-01-05 13:00:00   
5633               2129-01-05          2129-01-05 13:00:00   
5634               2129-01-05          2129-01-05 03:00:00   
5635               2129-01-05          2129-01-05 08:00:00   
5636               2129-01-05          2129-01-05 08:00:00   
5637               2129-01-04          2129-01-04 08:00:00   
5638               2129-01-03          2129-01-03 19:00:00   
5639               2129-01-03          2129-01-03 20:00:00   
5640               2129-01-03          2129-01-03 19:00:00   
5641               2129-01-05          2129-01-05 08:00:00   
5642               2129-01-03          2129-01-03 19:00:00   
5643               2129-01-05          2129-01-05 04:00:00   
5644               2129-01-05          2129-01-05 03:00:00   
5645               2129-01-03          2129-01-03 19:00:00   

     drug_exposure_end_date drug_exposure_end_datetime  drug_type_concept_id  \
5622             2129-01-05        2129-01-05 01:00:00                 32838   
5623             2129-01-04        2129-01-04 23:00:00                 32838   
5624             2129-01-05        2129-01-05 02:00:00                 32838   
5625             2129-01-04        2129-01-04 23:00:00                 32838   
5626             2129-01-05        2129-01-05 02:00:00                 32838   
5627             2129-01-05        2129-01-05 18:00:00                 32838   
5628             2129-01-10        2129-01-10 08:00:00                 32838   
5629             2129-01-11        2129-01-11 08:00:00                 32838   
5630             2129-01-08        2129-01-08 08:00:00        

In [9]:
patient1[patient1.dose_unit_source_value == 'TAB']
# 9 lignes, des exposure_id différentes, avec des dates différents (2129 ????)
# drug_concept_id being the name of the drug I guess, Quantity in Tablettes.


drug_exposure_id            person_id  drug_concept_id  \
5640 -4804305780328125843  3912882389848878631           723020   
5639 -1076779580276771548  3912882389848878631          1127433   
5623 -2528282390097198867  3912882389848878631         40232756   
5625  4557254971515693852  3912882389848878631         40232756   
5622  6996991851384754967  3912882389848878631         40232756   
5624  8598302712376844332  3912882389848878631         40232756   
5635 -6600566077470540957  3912882389848878631         21125875   
5636  8139119374042756604  3912882389848878631         19137328   
5641 -2002420539696298274  3912882389848878631         19077513   

     drug_exposure_start_date drug_exposure_start_datetime  \
5640               2129-01-03          2129-01-03 19:00:00   
5639               2129-01-03          2129-01-03 20:00:00   
5623               2129-01-04          2129-01-04 18:00:00   
5625               2129-01-04          2129-01-04 21:00:00   
5622               2129-01-05          2129-01-05 00:00:00   
5624               2129-01-05          2129-01-05 02:00:00   
5635               2129-01-05          2129-01-05 08:00:00   
5636               2129-01-05          2129-01-05 08:00:00   
5641               2129-01-05          2129-01-05 08:00:00   

     drug_exposure_end_date drug_exposure_end_datetime  drug_type_concept_id  \
5640             2129-01-05        2129-01-05 02:00:00                 32838   
5639             2129-01-05        2129-01-05 18:00:00                 32838   
5623             2129-01-04        2129-01-04 23:00:00                 32838   
5625             2129-01-04        2129-01-04 23:00:00                 32838   
5622             2129-01-05        2129-01-05 01:00:00                 32838   
5624             2129-01-05        2129-01-05 02:00:00                 32838   
5635             2129-01-05        2129-01-05 18:00:00                 32838   
5636             2129-01-05        2129-01-05 18:00:00                 32838   
5641             2129-01-05        2129-01-05 18:00:00                 32838   

      quantity  route_concept_id  visit_occurrence_id drug_source_value  \
5640       1.0           4167540  3761046692024108622       51079028620   
5639       2.0           4167540  3761046692024108622       51079000220   
5623       0.5           4167540  3761046692024108622       00406055262   
5625       1.0           4167540  3761046692024108622       00406055262   
5622       1.5           4167540  3761046692024108622       00406055262   
5624       1.0           4167540  3761046692024108622       00406055262   
5635       1.0           4167540  3761046692024108622       00904053061   
5636       1.0           4167540  3761046692024108622       00904054460   
5641       1.0           4167540  3761046692024108622       62584089701   

      drug_source_concept_id route_source_value dose_unit_source_value  
5640                45000972              PO/NG                    TAB  
5639                45325177              PO/NG                    TAB  
5623                45266765              PO/NG                    TAB  
5625                45266765              PO/NG                    TAB  
5622                45266765              PO/NG                    TAB  
5624                45266765              PO/NG                    TAB  
5635                45165542              PO/NG                    TAB  
5636                45114473              PO/NG                    TAB  
5641                44954855              PO/NG                    TAB

In [5]:
patient1_drug1 = patient1.loc[patient1.drug_concept_id==40232756,['drug_exposure_start_datetime', 'drug_exposure_end_datetime', 'quantity']].sort_values(by='drug_exposure_start_datetime')



patient1_drug1['cumulative'] = [0.5, 1.5, 3.0, 4.0]



# ne fonctionne pas comme ça, car les date ne sont pas correct.... il faut prendre datetime déjà, puis trier en fonction du temps.
# il faut faire avec plotly un graph pour choisir le medoc.
# il faut calculer le cumulatif ou non ?
patient1_drug1


drug_exposure_start_datetime drug_exposure_end_datetime  quantity  \
5623          2129-01-04 18:00:00        2129-01-04 23:00:00       0.5   
5625          2129-01-04 21:00:00        2129-01-04 23:00:00       1.0   
5622          2129-01-05 00:00:00        2129-01-05 01:00:00       1.5   
5624          2129-01-05 02:00:00        2129-01-05 02:00:00       1.0   

      cumulative  
5623         0.5  
5625         1.5  
5622         3.0  
5624         4.0

In [63]:
patient1_drug1.drug_exposure_start_datetime.min()

'2129-01-04 18:00:00'

In [25]:

fig = make_subplots(rows = 1, cols = 2, subplot_titles=('not cumulative','cumulative'))
fig.add_trace(go.Scatter(x = patient1_drug1.drug_exposure_start_datetime, y = patient1_drug1.quantity), row=1, col=1)
fig.update_layout(xaxis=dict(rangeslider=dict(autorange=True, range=[patient1_drug1.drug_exposure_start_datetime.min(), patient1_drug1.drug_exposure_start_datetime.max()])))
   
fig.add_trace(go.Scatter(x = patient1_drug1.drug_exposure_start_datetime, y = patient1_drug1.cumulative), row=1, col=2)
  
fig.show()

In [12]:
patient1 = patient1.sort_values(by='drug_exposure_start_datetime')
number_types = patient1.dose_unit_source_value.describe()['unique']

fig3 = make_subplots(rows = 1, cols = number_types, subplot_titles=(patient1.dose_unit_source_value.value_counts().index))
colonne = 1

for drug_type in patient1.dose_unit_source_value.value_counts().index:
    #print(drug_type)

    for d in patient1[patient1.dose_unit_source_value == drug_type].drug_concept_id.unique():
        temp = patient1[patient1.drug_concept_id == d]
        #print(temp.dose_unit_source_value.unique())
        fig3.add_trace(go.Scatter(x = temp.drug_exposure_start_datetime, y = temp.quantity, name=f'{d}, {temp.dose_unit_source_value.unique()}'), row=1, col=colonne)

    colonne += 1

fig3.show()

# Add curseur de temps, add legendgroup, add selection de chaque drug -> drug selection is already doable as well as time

In [13]:
# TOGGLE GRAPH - not useful

patient1 = patient1.sort_values(by='drug_exposure_start_datetime')
number_types = patient1.dose_unit_source_value.describe()['unique']

#fig3 = make_subplots(rows = 1, cols = number_types, subplot_titles=(patient1.dose_unit_source_value.value_counts().index))
colonne = 1

for drug_type in patient1.dose_unit_source_value.value_counts().index:
    #print(drug_type)
    traces = []
    buttons = []

    # for each drug type, create a liste of traces and buttons
    for d in patient1[patient1.dose_unit_source_value == drug_type].drug_concept_id.unique():
        temp = patient1[patient1.drug_concept_id == d]

        traces.append(go.Scatter(x = temp.drug_exposure_start_datetime, y = temp.quantity, name=f'{d}, {temp.dose_unit_source_value.unique()}', visible = True))
        
        buttons.append(dict(method='update',
                        label=f'{d}',
                        visible=True,
                        args=[{'visible':[x == d for x in patient1[patient1.dose_unit_source_value == drug_type].drug_concept_id.unique()]}],
                        args2=[{'visible':[x != d for x in patient1[patient1.dose_unit_source_value == drug_type].drug_concept_id.unique()]}]
                        )
                )

    # Make the graph with toggle
        
    # add a button 'All' to toggle them all together    
    layout = go.Layout(
        updatemenus=[
            dict(
                type='buttons',
                direction='right',
                x=0.7,
                y=1.3,
                showactive=True,
                buttons=buttons
            )
        ],
        showlegend=True

    )

    fig3 = go.Figure(data=traces,layout=layout)
    fig3.update_layout(xaxis=dict(rangeslider=dict(autorange=True, range=[patient1_drug1.drug_exposure_start_datetime.min(), patient1_drug1.drug_exposure_start_datetime.max()])))
   
    fig3.show()




# https://plotly.com/python/range-slider/ axes avec plusieurs type d'échelle

### Drug_exposure start and end can be different

In [ ]:
# drug_exposure.drug_exposure_start_datetime.value_counts()
#drug_exposure.drug_exposure_end_datetime.value_counts()
print(drug_exposure.drug_exposure_start_datetime[0], drug_exposure.drug_exposure_end_datetime[0])
# Can be different....

# Dose Era

A Dose Era is defined as a span of time when the Person is assumed to be exposed to a constant dose of a specific active ingredient.

Dose Form information is not taken into account. So, if the patient changes between different formulations, or different manufacturers with the same formulation, the Dose Era is still spanning the entire time of exposure to the Ingredient.

In [15]:
dose_era = pd.read_csv('Data/dose_era.csv')
dose_era.head()

#dose_era.shape #(117, 7)

dose_era_id            person_id  drug_concept_id  unit_concept_id  \
0 -1101936257632052351  5548892236933978704           902427             8576   
1  9002582707624403415  4985579811051920670           714785             8576   
2  5495900697391314629 -4873075614181207858          1328165             8576   
3  6335084700122752427 -8205283012979532608          1328165             8576   
4  2848105832570024513 -4873075614181207858          1328165             8576   

   dose_value dose_era_start_date dose_era_end_date  
0        10.0          2142-05-14        2142-05-16  
1         2.5          2110-04-13        2110-04-14  
2       120.0          2148-09-15        2148-09-15  
3       120.0          2189-06-11        2189-06-13  
4       120.0          2147-12-20        2147-12-21

In [13]:
dose_era.unit_concept_id.value_counts()
# 8576 are all in mg.... This is good :) I can go back to drug_concept id, dose_value....

unit_concept_id
8576    117
Name: count, dtype: int64

In [16]:
dose_era.drug_concept_id.value_counts()

drug_concept_id
1307046    38
1328165    12
718583     12
1383815    11
745466     10
965748      8
750982      7
1398937     4
1503297     2
1560171     2
1163944     2
1318853     2
1154029     2
714785      1
1353776     1
719311      1
1353766     1
902427      1
Name: count, dtype: int64

# Drug era

**active ingredient**, successive period of drug_exposure to the same ingredent gives drug era

A Drug Era is defined as a span of time when the Person is assumed to be exposed to a particular active ingredient. A Drug Era is not the same as a Drug Exposure:
Exposures are individual records corresponding to the source when Drug was delivered to the Person, while successive periods of Drug Exposures are combined under certain rules to produce continuous Drug Eras.

-> not the same concept_id because here we talk about an ingredient, not a particular drug... diazepam vs diazepam tablet 10mg.  

Example:  
drug_era: person 3912882389848878631 got 2 exposure to diazepam 723013 between 2129-01-03	2129-01-05  
[6598	-7540355434127132424	3912882389848878631	723013	2129-01-03	2129-01-05	2	0]   

drug_exposure: person 3912882389848878631 got 1 drug_exposure to 723020 within 2129-01-03 19:00:00	and 2129-01-05 02:00:00	  
[5640	-4804305780328125843	3912882389848878631	723020	2129-01-03	2129-01-03 19:00:00	2129-01-05	2129-01-05 02:00:00	NaN	32838	NaN	...	NaN	4167540	NaN	NaN	3761046692024108622	NaN	51079028620	45000972	PO/NG	TAB]

In [14]:
dose_era[dose_era.person_id == 3912882389848878631] # my patient is not recognized, neither my drug_cncept_id...why ????
#patient1[patient1.dose_unit_source_value == drug_type]
dose_era[dose_era.drug_concept_id == 19127213]
# ne pas mettre de guillements, ça ne fonctionne pas...

Empty DataFrame
Columns: [dose_era_id, person_id, drug_concept_id, unit_concept_id, dose_value, dose_era_start_date, dose_era_end_date]
Index: []

In [4]:
drug_era = pd.read_csv('Data/drug_era.csv')

drug_era.shape #(7931, 7)
drug_era.head()

drug_era_id            person_id  drug_concept_id  \
0 -4731852071843461961 -2067961723109232727          1124957   
1  1725818565432188772  -626229666378242477          1368671   
2  3222075720408667327  -626229666378242477           967823   
3  7411613425123823172   579254014084392336          1310149   
4  4265287408545791371 -2312013739856114142         19095164   

  drug_era_start_date drug_era_end_date  drug_exposure_count  gap_days  
0          2169-01-21        2169-01-24                    1         0  
1          2171-11-11        2171-11-16                    2         1  
2          2171-11-11        2171-11-22                   12         0  
3          2115-10-16        2115-10-18                    2         0  
4          2116-06-28        2116-07-05                    1         0

In [ ]:
drug_era.drug_concept_id.value_counts()

In [5]:
drug_era[drug_era.drug_concept_id == 19127213] # cannot be found...

drug_era[drug_era.person_id == 3912882389848878631] # -> ok...

drug_era_id            person_id  drug_concept_id  \
6596   500595522397761829  3912882389848878631         43531944   
6597  6330107922049835930  3912882389848878631          1125315   
6598 -7540355434127132424  3912882389848878631           723013   
6599  7165454045076608944  3912882389848878631         40166605   
6600 -8180642557726588733  3912882389848878631         40164828   
6601  3025930947898329035  3912882389848878631         42873956   
6602  8800459384878548729  3912882389848878631           967823   
6603  -338966274758502286  3912882389848878631          1367571   
6683  1590324940605174808  3912882389848878631           923645   
6684  8866077140124362516  3912882389848878631          1124957   
6690 -1112804010670705686  3912882389848878631         36878782   
6691 -2016335706505619184  3912882389848878631         19111620   
6692  2426943909930076013  3912882389848878631          1560524   
6693  8939824964893766208  3912882389848878631         19137312   
6694  4416439696488798469  3912882389848878631          1177480   
6695  -945752478111843518  3912882389848878631           734275   

     drug_era_start_date drug_era_end_date  drug_exposure_count  gap_days  
6596          2129-01-03        2129-01-05                    1         0  
6597          2129-01-03        2129-01-05                    1         0  
6598          2129-01-03        2129-01-05                    2         0  
6599          2129-01-03        2129-01-05                    1         0  
6600          2129-01-03        2129-01-05                    1         0  
6601          2129-01-03        2129-01-05                    1         0  
6602          2129-01-03        2129-01-05                    2         0  
6603          2129-01-03        2129-01-05                    1         0  
6683          2129-01-04        2129-01-05                    1         0  
6684          2129-01-04        2129-01-05                    4         1  
6690          2129-01-05        2129-01-05                    1         0  
6691          2129-01-05        2129-01-05                    1         0  
6692          2129-01-05        2129-01-05                    1         0  
6693          2129-01-05        2129-01-05                    1         0  
6694          2129-01-05        2129-01-05                    1         0  
6695          2129-01-05        2129-01-11                    7         6

In [ ]:
temp_1 = drug_era.loc[drug_era.person_id == 3912882389848878631, 'drug_concept_id'].unique() #16
temp_1.sort()
'''
drug_era
(array([  723013 - 	diazepam as an ingredient,   734275,   923645,   967823,  1124957,  1125315,
         1177480,  1367571,  1560524, 19111620, 19137312, 36878782,
        40164828, 40166605, 42873956, 43531944]),

drug_exposure
 array([  723020 - diazepam 10 MG Oral Tablet RxNorm,   734281,  1127433, 19019050, 19019418, 19077513,
        19079250, 19126251, 19127213, 19137328, 21125875, 36249736,
        40221329, 40232756, 43011850, 45775137]))
'''
temp_2 = drug_exposure.loc[drug_exposure.person_id == 3912882389848878631, 'drug_concept_id'].unique() #16
temp_2.sort()

# all the same

temp_1, temp_2


In [10]:
temp_3 = dose_era.loc[dose_era.person_id == 3912882389848878631]
temp_3 # none
#temp_3.loc[temp_3.drug_concept_id == 723013]

Empty DataFrame
Columns: [dose_era_id, person_id, drug_concept_id, unit_concept_id, dose_value, dose_era_start_date, dose_era_end_date]
Index: []

In [ ]:
temp_exp = drug_exposure.person_id.unique()
len(temp_exp) #100
temp_drug = drug_era.person_id.unique()
len(temp_drug) #100
temp_dose = dose_era.person_id.unique()
len(temp_dose) #44

In [17]:
drug_strength = pd.read_csv('Data/DRUG_STRENGTH.csv', sep='\s+')
drug_strength.shape #(200983, 12)

'''
Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.


drug_concept_id                  int64
ingredient_concept_id            int64
amount_value                   float64
amount_unit_concept_id           int64
numerator_value                float64
numerator_unit_concept_id       object
denominator_value              float64
denominator_unit_concept_id    float64
box_size                       float64
valid_start_date               float64
valid_end_date                 float64
invalid_reason                 float64
'''

drug_strength.head()

/tmp/ipykernel_5721/1970751477.py:1: DtypeWarning:

Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.



drug_concept_id  ingredient_concept_id  amount_value  \
0          1154535                1154343         0.417   
1          1154536                1154343         0.210   
2          1731598                1731597         2.000   
3          1396092                1396012       225.000   
4          1517764                1517740         0.002   

   amount_unit_concept_id  numerator_value numerator_unit_concept_id  \
0                    8576           8587.0                  19700101   
1                    8576           8587.0                  19700101   
2                    8576           8587.0                  19700101   
3                    8576       19700101.0                  20991231   
4                    8576           8587.0                  19700101   

   denominator_value  denominator_unit_concept_id  box_size  valid_start_date  \
0         20991231.0                          NaN       NaN               NaN   
1         20991231.0                          NaN       NaN               NaN   
2         20991231.0                          NaN       NaN               NaN   
3                NaN                          NaN       NaN               NaN   
4         20991231.0                          NaN       NaN               NaN   

   valid_end_date  invalid_reason  
0             NaN             NaN  
1             NaN             NaN  
2             NaN             NaN  
3             NaN             NaN  
4             NaN             NaN

In [27]:
drug_strength.info()
'''
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200983 entries, 0 to 200982
Data columns (total 12 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   drug_concept_id              200983 non-null  int64  
 1   ingredient_concept_id        200983 non-null  int64  
 2   amount_value                 200983 non-null  float64
 3   amount_unit_concept_id       200983 non-null  int64  
 4   numerator_value              200983 non-null  float64
 5   numerator_unit_concept_id    142454 non-null  object 
 6   denominator_value            79593 non-null   float64
 7   denominator_unit_concept_id  7763 non-null    float64
 8   box_size                     0 non-null       float64
 9   valid_start_date             0 non-null       float64
 10  valid_end_date               0 non-null       float64
 11  invalid_reason               0 non-null       float64
dtypes: float64(8), int64(3), object(1)
memory usage: 18.4+ MB
'''


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200983 entries, 0 to 200982
Data columns (total 12 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   drug_concept_id              200983 non-null  int64  
 1   ingredient_concept_id        200983 non-null  int64  
 2   amount_value                 200983 non-null  float64
 3   amount_unit_concept_id       200983 non-null  int64  
 4   numerator_value              200983 non-null  float64
 5   numerator_unit_concept_id    142454 non-null  object 
 6   denominator_value            79593 non-null   float64
 7   denominator_unit_concept_id  7763 non-null    float64
 8   box_size                     0 non-null       float64
 9   valid_start_date             0 non-null       float64
 10  valid_end_date               0 non-null       float64
 11  invalid_reason               0 non-null       float64
dtypes: float64(8), int64(3), object(1)
memory usage: 18.4+ MB


In [29]:
drug_strength[drug_strength['drug_concept_id']==734281] # it works
'''	
phenobarbital 65 MG/ML

drug_concept_id	ingredient_concept_id	amount_value	amount_unit_concept_id	numerator_value	numerator_unit_concept_id	denominator_value	denominator_unit_concept_id	box_size	valid_start_date	valid_end_date	invalid_reason
734281	        734275	                65.0	            8576	                8587.0	    19700101	                20991231.0	            NaN	NaN	NaN	NaN	NaN
8576 = mg... should be mg/ml
8587 = mL...numeratot ! That's why. check with a TAB and others, check the numerator_unit_conceptid
'''

drug_concept_id  ingredient_concept_id  amount_value  \
38328           734281                 734275          65.0   

       amount_unit_concept_id  numerator_value numerator_unit_concept_id  \
38328                    8576           8587.0                  19700101   

       denominator_value  denominator_unit_concept_id  box_size  \
38328         20991231.0                          NaN       NaN   

       valid_start_date  valid_end_date  invalid_reason  
38328               NaN             NaN             NaN

In [32]:
drug_strength.numerator_value.unique() 
drug_strength.amount_unit_concept_id.unique()

array([    8576,     8510,     9551, 45744811,     8587,     9573,
       45744812, 45744810,     9693, 44819154, 45744813,    32018,
          32407, 45744814,     9278,     9439,     9367, 45744816,
           8718,     8554, 19700101, 20050911, 20080302, 20080330,
       20080629, 20080831, 20080928, 20090201, 20220103, 20070128,
       20080427, 20070429, 20060514, 20060226, 20071230, 20050724,
       20210405, 20230103, 20220801, 20060709, 20051204, 20070729,
       20060402, 20190603, 20180205, 20060115, 20070225, 20051113,
       20080601, 20080127, 20180402, 20070527, 20070701, 20200803,
       20060312, 20070902, 20070401, 20060730, 20090104, 20190204,
       20060827, 20070114, 20190903, 20190401, 20190701, 20180702,
       20190506, 20221003, 20210802, 20180806, 20061119, 20061217,
       20170807, 20170103, 20220307, 20070930, 20090503, 20081130,
       20090531, 20080727, 20081109, 20060604, 20090301, 20061008,
       20181203, 20160502, 20210503, 20160801, 20210104, 20200